Src/Newtonsoft.Json/JsonTextWriterInternal.cs

```diff
@@ -0,0 +1,356 @@
+﻿#region License
+// Copyright (c) 2007 James Newton-King
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+#endregion
+
+using System.IO;
+#if !(NET20 || NET35 || NET40 || PORTABLE40)
+using System;
+#if !PORTABLE || NETSTANDARD1_1
+using System.Numerics;
+#endif
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Utilities;
+#endif
+
+namespace Newtonsoft.Json
+{
+    // A sealed class derived from JsonTextWriter that, as it's internal and sealed, we know isn't
```

Comment:
Right now it'll be marginal, but I introduced it in my first stab at this since async serialisation would hit it many times.
OTOH, the caching of _safeAsync has made that check a lot cheaper. Once I've no longer got a toddler asleep on top of me I'll take it out. The benefits can be examined again if async serialisation means there's a path that does more hits on it again.